client/state/data-layer/wpcom-http/pipeline/retry-on-failure/README.md

```diff
@@ -0,0 +1,115 @@
+# Retry network requests on failure
+
+In 1985 Jim Gray proposed the following conjecture<sup>1</sup>…
+
+> most production…faults are soft. If…reinitialized and…retried, the operation will usually not fail the second time.
+
+This suggests that there are a large variety of bugs in software due to inexplicable and ephemeral causes.
+Many of our network connections experience these types of _Heisenbugs_ which occur at random and are hard to debug.
+This pipeline processor exists to limit the extent to which these temporary failures impact Calypso and those developers writing code which interacts with HTTP requests.
+
+## What is a failed network request?
+
+There are two reasons a network request can fail.
+The first is because the request was invalid and the remote server legitimately returned an error.
+The second is because something along the network path failed even though the request was legitimate.
+Network failures can occur on slow network connections, when transitioning from one network to another, when physical cables are cut, when routers crash, and in one of any number of different ways.
+
+Because most of these failures are transient we can pause before considering a failed request as a true failure.
+A _failed request_ is one which fails after we have given it every reasonable attempt to succeed.
+
+## What happens with failing requests?
+
+We really don't want to have to write separate error handlers for these kinds of transient faults in every bit of code which interacts with HTTP requests.
+Instead, we can apply _policies_ to the request descriptions which indicate how failures should be treated and provide a default behavior.
+When the request comes back as a failure we track its state and simply resend the request (if we determine that we can).
+Repeated failures are compared against the state tracking to determine if we should continue to attempt resending it.
+Once the _reasonable effort_ has been exhausted we consider the request truly as a failure and reinject it to the HTTP layer where the `onFailure` action will be dispatched.
+
+### How can we safely replay a failed network request?
+
+Certain requests are assumed to be _safe_ and _idempotent_ if they are `GET` requests with matching `path`, version, and `query` parameters.
+
+## Failure modes and policies
+<!-- diagram made in draw.io and source XML found at https://cldup.com/H2-4aSzJup.xml -->
+
+### No retry
+![](https://cldup.com/SlUCoIuzkX.svg)
+
+This policy means that once a request has failed we should _not_ attempt to retry it.
+Instead, the request should immediately fail.
+This should be used for requests which Calypso might think are safe to resend but which aren't.
+In the diagram we can see that the initial request failed and then immediately the HTTP layer dispatched the `onFailure` action.
+
+### Simple retry
+![](https://cldup.com/FQFuTX5CFq.svg)
+
+Please don't use this policy unless you have a very good reason not to use the exponential-backoff.
```

Comment:
ah but @gwwar can you think of a reason this would be a better choice? in most cases the early retries will be close to identical.